### ISSUE TITLE
B::MAGIC::MOREMAGIC returns bad B::MAGIC if no more magic

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5876]
Migrated from [rt.perl.org#16828](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=16828) (status was 'resolved')

Searchable as RT16828$

**Comment by p5pRT** [comment]
### From @ysth

#### Created by @ysth



When you reach the end of the magic chain\, MOREMAGIC returns a
B&#8203;::MAGIC obj containing a null pointer\.  Using it then segfaults\.
MOREMAGIC should probably return undef instead\.

~ $perl \-MB \-we'B&#8203;::svref\_2object\(\\$\.\)\->MAGIC\->MOREMAGIC\->TYPE'

Segmentation fault \(core dumped\)

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.0:

Configured by sthoenna at Sun Aug  4 15:00:09  2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=cygwin, osvers=1.3.10(0.5132), archname=cygwin-64int
    uname='cygwin_98-4.10 yo 1.3.10(0.5132) 2002-02-25 11:14 i686 unknown '
    config_args='-de -Dusedevel -Duse64bitint -Ui_ndbm -Ui_dbm -Dstatic_ext='IO Socket' -Doptimize=-g'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=define use64bitall=undef uselongdouble=undef
    usemymalloc=y, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-DPERL_USE_SAFE_PUTENV -DDEBUGGING -fno-strict-aliasing -I/usr/local/include',
    optimize='-g',
    cppflags='-DPERL_USE_SAFE_PUTENV -DDEBUGGING -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.3-5 (cygwin special)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='ld2', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib /lib
    libs=-lgdbm -lcrypt -lutil
    perllibs=-lcrypt -lutil
    libc=/usr/lib/libc.a, so=dll, useshrplib=true, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags=' -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /usr/local/lib/perl5/5.8.0/cygwin-64int
    /usr/local/lib/perl5/5.8.0
    /usr/local/lib/perl5/site_perl/5.8.0/cygwin-64int
    /usr/local/lib/perl5/site_perl/5.8.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.8.0:
    HOME=/cygdrive/c/home/sthoenna
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/usr/bin:/bin:/cygdrive/c/home/sthoenna/bin:/cygdrive/c/PROGRA~1/YARNW:/usr/bin:/USR/LOCAL/EMACS/EMACS-20.7/BIN:/cygdrive/c/WINDOWS:/cygdrive/c/WINDOWS/COMMAND
    PERL_BADLANG (unset)
    SHELL (unset)

```
</details>

**Comment by p5pRT** [comment]
### From @lizmat

At 01&#8203;:47 AM 8/29/02 \+0000\, via RT wrote&#8203;:
>~ $perl \-MB \-we'B&#8203;::svref\_2object\(\\$\.\)\->MAGIC\->MOREMAGIC\->TYPE'
>Segmentation fault \(core dumped\)

For what it's worth\, the associated valgrind output&#8203;:

==1245== Invalid read of size 1
==1245==    at 0x43583950&#8203;: XS\_B\_\_MAGIC\_TYPE \(in 
/usr/local/lib/perl5/5\.8\.0/i686\-linux\-thread\-multi/auto/B/B\.so\)
==1245==    by 0x80DC4D8&#8203;: Perl\_pp\_entersub \(pp\_hot\.c&#8203;:2773\)
==1245==    by 0x80BA60B&#8203;: Perl\_runops\_debug \(dump\.c&#8203;:1398\)
==1245==    by 0x80643DA&#8203;: S\_run\_body \(perl\.c&#8203;:1681\)
==1245==    by 0x8063E53&#8203;: perl\_run \(perl\.c&#8203;:1600\)
==1245==    by 0x805F4C9&#8203;: main \(perlmain\.c&#8203;:85\)
==1245==    by 0x402B4507&#8203;: \_\_libc\_start\_main 
\(\.\./sysdeps/generic/libc\-start\.c&#8203;:129\)
==1245==    by 0x805F321&#8203;: strcpy@&#8203;@&#8203;GLIBC\_2\.0 \(in /usr/bin/perl5\.8\.0\-debug\)
==1245==    Address 0xA is not stack'd\, malloc'd or free'd


Liz

**Comment by p5pRT** [comment]
### From @rgs

Yitzchak Scott\-Thoennes \(via RT\) wrote&#8203;:
> When you reach the end of the magic chain\, MOREMAGIC returns a
> B&#8203;::MAGIC obj containing a null pointer\.  Using it then segfaults\.
> MOREMAGIC should probably return undef instead\.

This should cure it :

Change 17823 by rgs@&#8203;rgs\-home on 2002/09/03 19&#8203;:47&#8203;:05

&nbsp;    Fix bug \#16828\.
&nbsp;    Add a few tests to ext/B/t/b\.t\, make it use Test&#8203;::More\.

Affected files \.\.\.

\.\.\.\.\.\. //depot/perl/ext/B/B\.xs\#77 edit
\.\.\.\.\.\. //depot/perl/ext/B/t/b\.t\#2 edit

Differences \.\.\.

==== //depot/perl/ext/B/B\.xs\#77 \(text\) ====

@&#8203;@&#8203; \-1077\,6 \+1077\,15 @&#8203;@&#8203;
 B&#8203;::MAGIC
 MgMOREMAGIC\(mg\)
&nbsp;     B&#8203;::MAGIC    mg
\+     CODE&#8203;:
\+    if\( MgMOREMAGIC\(mg\) \) \{
\+        RETVAL = MgMOREMAGIC\(mg\);
\+    \}
\+    else \{
\+        XSRETURN\_UNDEF;
\+    \}
\+     OUTPUT&#8203;:
\+    RETVAL
 
 U16
 MgPRIVATE\(mg\)

==== //depot/perl/ext/B/t/b\.t\#2 \(xtext\) ====

@&#8203;@&#8203; \-13\,15 \+13\,9 @&#8203;@&#8203;
 $|  = 1;
 use warnings;
 use strict;
\-use Config;
\+use Test&#8203;::More tests => 5;
 
\-print "1\.\.2\\n";
\-
\-my $test = 1;
\-
\-sub ok \{ print "ok $test\\n"; $test\+\+ \}
\-
\-use B;
\+BEGIN \{ use\_ok\( 'B' \); \}
 
 
 package Testing&#8203;::Symtable;
@&#8203;@&#8203; \-55\,9 \+49\,18 @&#8203;@&#8203;
 push @&#8203;syms\, "Testing&#8203;::Symtable&#8203;::Foo&#8203;::yarrow";
 
 \# Make sure we hit all the expected symbols\.
\-print "not " unless join\(''\, sort @&#8203;syms\) eq join\(''\, sort keys %Subs\);
\-ok;
\+ok\( join\(''\, sort @&#8203;syms\) eq join\(''\, sort keys %Subs\)\, 'all symbols found' \);
 
 \# Make sure we only hit them each once\.
\-print "not " unless \!grep $\_ \!= 1\, values %Subs;
\-ok;
\+ok\( \(\!grep $\_ \!= 1\, values %Subs\)\, '\.\.\.and found once' \);
\+
\+\# Tests for MAGIC / MOREMAGIC
\+ok\( B&#8203;::svref\_2object\(\\$\.\)\->MAGIC\->TYPE eq "\\0"\, '$\. has \\0 magic' \);
\+\{
\+    my $e = '';
\+    local $SIG\{\_\_DIE\_\_\} = sub \{ $e = $\_\[0\] \};
\+    \# Used to dump core\, bug \#16828
\+    eval \{ B&#8203;::svref\_2object\(\\$\.\)\->MAGIC\->MOREMAGIC\->TYPE; \};
\+    like\( $e\, qr/Can't call method "TYPE" on an undefined value/\, 
\+    '$\. has no more magic' \);
\+\}

**Comment by p5pRT** [comment]
@rspier - Status changed from 'new' to 'resolved'
